### PR TITLE
Static check noalloc

### PR DESCRIPTION
**Pull request by gretay-js**
This is the first part of #707 :
- only allocation check
- only backend, command-line flag -check-alloc and cmx file format changes
- no attributes
- all raises are allowed (raise with trace is not treated as allocating)

### CODE REVIEW

**Comment by mshinwell** [comment]
@xclerc is going to review the parts under the `backend` directory, I'll review the remainder.

File: backend/checkmach.ml [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+
+  let check_callee s (checks : Cmx_format.checks) =
+    StringSet.mem s checks.ui_noalloc_functions
+
+  let add_callee s (checks : Cmx_format.checks) =
+    if not (check_callee s checks)
+    then
+      checks.ui_noalloc_functions <- StringSet.add s checks.ui_noalloc_functions
+
+  (** conservative *)
```
Comment (xclerc) [inline backend/checkmach.ml:424]: I am not sure it is; do the "checkbound"
(arm64) instructions allocate? Even if
they don't, the check can be silently
broken by the addition of a new operation.

**Review by xclerc — COMMENTED** [review]
Just made a first pass (would like to read
`Checkmach` again if we are not in a hurry.)

Two global questions:

- are we sure we want the checks to always be
  hard errors, or are there workflows where
  treating them more like warnings would make
  sense?
- is it difficult to adds some tests? (I don't know
  ocamltest enough to be sure, but it feels like
  we could check the basics through simple
  compiler invocations.)

File: backend/checkmach.ml (hunk 1 above)
Comment (gretay-js) [reply backend/checkmach.ml:424]: My reasoning was if an operation allocates, it must be marked as "can_raise".

File: backend/printcmm.ml [hunk 2]
```diff
@@ -339,6 +339,15 @@ and sequence ppf = function
 
 and expression ppf e = fprintf ppf "%a" expr e
 
+let print_codegen_option = function
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/printcmm.ml]: I changed name to `codegen_option`

File: middle_end/compilenv.ml [hunk 3]
```diff
@@ -54,6 +54,7 @@ module CstMap =
   end)
 
 module SymMap = Misc.Stdlib.String.Map
+module S = Misc.Stdlib.String.Set
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply middle_end/compilenv.ml]: Done.

File: middle_end/compilenv.mli [hunk 4]
```diff
@@ -112,6 +112,25 @@ val need_send_fun: int -> Lambda.alloc_mode -> unit
...
 
+module Checks : sig
+  (* mutable state *)
+  type t = Cmx_format.checks
+
+  val create : unit -> t
+
+  val reset : t -> unit
+
+  (* [merge_checks c ~into] modifies [into] by adding information from [src]. *)
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply middle_end/compilenv.mli]: Fixed.

File: file_formats/cmx_format.mli [hunk 5]
```diff
@@ -44,6 +44,13 @@ type generic_fns =
     apply_fun: apply_fn list;
     send_fun: apply_fn list }
 
+(* Symbols of function that pass certain checks for special properties. *)
+type checks =
+  {
+    mutable ui_noalloc_functions: Misc.Stdlib.String.Set.t;
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply file_formats/cmx_format.mli:51]: I added a CR. It is used directing in `checkmach` for membership tests. I'll do a separate refactoring of `checkmach` to use a set internally and at the end to convert it to a list.

**Review by gretay-js — COMMENTED** [review]
@xclerc I've addressed the comments.

File: backend/checkmach.ml [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+    | Fail -> "fail"
+    | Unknown -> "unknown"
+end
+
+module Func_info = struct
+  type t =
+    { name : string;  (** function name *)
+      mutable value : Value.t;  (** the result of the check *)
+      mutable callers : string list;
+          (** unresolved dependencies. if not empty then value is Unknown. *)
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/checkmach.ml:55]: (Discussed off-line; it is probably slightly
more convenient the way it is right now.)

File: backend/checkmach.ml [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+
+  let create name value =
+    { name; value; callers = []; in_current_unit = false; annotated = false }
+end
+
+(* Return true if [str] starts with "caml_apply". This is conservative. *)
+let is_caml_apply str =
+  (* CR-someday gyorsh: propagate information about caml_apply from cmm to mach
+     instead of reverse-engineering it from the symbol name. *)
+  let caml_apply_prefix = "caml_apply" in
```
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/checkmach.ml:69]: Sorry, I meant: it looks like you are
re-implementing `String.begins_with`;
the name of the function was fine.

File: backend/checkmach.ml (hunk 1 above)
Comment (gretay-js) [reply backend/checkmach.ml:424]: I don't know why. My reasoning was based on if Ialloc can raise than anything else that allocations should also be marked as can_raise.

Comment (xclerc) [reply backend/checkmach.ml:424]: Is it something that could somehow be affected
by polling points? (I am out of my depth.)

File: backend/checkmach.ml (hunk 7 above)
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/checkmach.ml]: Fixed, thanks! (I didn't know we have this function.)

[the remark this answers is not in the document]
Comment (xclerc) [reply backend/checkmach.ml]: (Me neither, I think I have reimplemented it more than once...)

**Comment by gretay-js** [comment]
> * are we sure we want the checks to always be
>   hard errors, or are there workflows where
>   treating them more like warnings would make
>   sense?

There are some internal consistency checks that fail with `fatal_error`, but you are probably referring to the error raise if function doesn't confirm to its noalloc annotations (that are implemented in a separate PR). First of all, nothing is annotated to start with, and the noalloc check is off by default in user code (on by default in the compiler distribution). When annotations are added, if a change in a callee causes an "@assert noalloc" annotation to be violated, then either the callee needs to be fixed, or annotated with "@assume noalloc". If this is not enough, we can add some way of turning them into warnings. I first wanted to make it a warning not an error, so that can be turned into a hard error using -warn-error, but Warnings is not used in th backend and it didn't seem right to add one.

File: backend/checkmach.ml (hunk 1 above)
Comment (mshinwell) [reply backend/checkmach.ml:420]: Those will also need to be marked as raising, until we merge the fix for asynchronous exceptions.  At that point neither allocation points nor polling points will raise.

Comment (xclerc) [reply backend/checkmach.ml:420]: Discussed off-line: we should just add a
dedicated function for allocation, even if
for now its semantics is aligned with the
one about raises.

File: backend/amd64/arch.ml [hunk 8]
```diff
@@ -263,6 +263,8 @@ let operation_can_raise = function
   | Istore_int (_, _, _) | Ioffset_loc (_, _)
   | Iprefetch _ -> false
 
+let operation_allocates _ = false
```
Comment (xclerc) [inline backend/amd64/arch.ml:266]: I know I am being annoying, but using
a catch-all pattern essentially defeats
the purpose of the new function: the
compiler will not warn us when we add
new instructions.

Comment (gretay-js) [reply backend/amd64/arch.ml:266]: good point, thanks! fixing

**Review by xclerc — APPROVED** [review]

**Review by mshinwell — APPROVED** [review]

File: ocaml/otherlibs/bigarray/dune [hunk 9]
```diff
@@ -20,7 +20,7 @@
    -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66
    -warn-error A -bin-annot -safe-string -strict-formats
  ))
- (ocamlopt_flags (:include %{project_root}/ocamlopt_flags.sexp))
+ (ocamlopt_flags (:include %{project_root}/ocamlopt_flags.sexp) -alloc-check)
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply ocaml/otherlibs/bigarray/dune]: Discussed offline

**Pull request merged**